# Chapter 11

## Assumptions for this chapter
1. {$x_{t1}, x_{t2}, ..., x_{tk}, y_t$, $t = 1, 2, ..., n$} is stationary, weakly dependent and follows the linear model $y_t = \beta_0 + \beta_1 x_{t1} + \beta_2 x_{t2} + ... \beta_k x_{tk} + u_t$ (where $u_t$ is the sequence of errors).
2. No perfect collinearity.
3. The explanatory variables are contemporaneously exogenous ($E(u_t | x_{t1}, ..., x_{tk}) = 0$). This is different than the previous chapter where exogenity was required for all periods.
4. The errors are contemporaneously homoskedastic ($Var(u_t | (x_{t1}, x_{t2}, ..., x_{tk})) = \sigma^2$). Again, contemporary instead of over all time periods.
5. No serial correlation (For all $t \neq s, E(u_t u_s | x_t, x_s) = 0$).

Normality is no longer assumed in this chapter. Weak dependence (correlation between $x_t, x_{t+h}$ go to $0$ "sufficiently quickly" as $h$ increases withou bound) is what we need for this to work.

In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy as np

In [2]:
# Exercise 1
hseinv = pd.read_stata("stata/HSEINV.DTA")
hseinv.linvpc.corr(hseinv.linvpc.shift(1))

0.639124587909112

In [3]:
# We know from chapter 10 that we can use the residuals from a regression on the time trend to detrend
resids = sm.OLS(hseinv.linvpc, sm.add_constant(hseinv.t)).fit().resid
resids.corr(resids.shift(1))

0.4847401433757272

In [4]:
hseinv.lprice.corr(hseinv.lprice.shift(1))

0.9491585505000759

In [5]:
resids = sm.OLS(hseinv.lprice, sm.add_constant(hseinv.t)).fit().resid
resids.corr(resids.shift(1))

0.8215255620225328

In [6]:
X = sm.add_constant(hseinv[["gprice", "t"]])
sm.OLS(hseinv.linvpc, X, missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 linvpc   R-squared:                       0.510
Model:                            OLS   Adj. R-squared:                  0.484
Method:                 Least Squares   F-statistic:                     19.77
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.31e-06
Time:                        06:03:16   Log-Likelihood:                 30.090
No. Observations:                  41   AIC:                            -54.18
Df Residuals:                      38   BIC:                            -49.04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.8533      0.040    -21.177      0.000      -0.935      -0.772
gprice         3.8786      0.958      4.049      0.000       1.939       5.818
t              0.0080      0.002      5.038      0.000       0.005       0.011
==============================================================================
Omnibus:                        2.131   Durbin-Watson:                   0.930
Prob(Omnibus):                  0.345   Jarque-Bera (JB):                1.137
Skew:                          -0.300   Prob(JB):                        0.566
Kurtosis:                       3.553   Cond. No.                     1.27e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.27e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [7]:
hseinv["linvpc_dt"] = sm.OLS(hseinv.linvpc, sm.add_constant(hseinv.t)).fit().resid
sm.OLS(hseinv.linvpc_dt, X, missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              linvpc_dt   R-squared:                       0.303
Model:                            OLS   Adj. R-squared:                  0.266
Method:                 Least Squares   F-statistic:                     8.242
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00106
Time:                        06:03:16   Log-Likelihood:                 30.090
No. Observations:                  41   AIC:                            -54.18
Df Residuals:                      38   BIC:                            -49.04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0120      0.040     -0.297      0.768      -0.094       0.070
gprice         3.8786      0.958      4.049      0.000       1.939       5.818
t             -0.0001      0.002     -0.068      0.946      -0.003       0.003
==============================================================================
Omnibus:                        2.131   Durbin-Watson:                   0.930
Prob(Omnibus):                  0.345   Jarque-Bera (JB):                1.137
Skew:                          -0.300   Prob(JB):                        0.566
Kurtosis:                       3.553   Cond. No.                     1.27e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.27e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [8]:
X = sm.add_constant(hseinv[["gprice", "t"]])
sm.OLS(hseinv.ginvpc, X, missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 ginvpc   R-squared:                       0.047
Model:                            OLS   Adj. R-squared:                 -0.003
Method:                 Least Squares   F-statistic:                    0.9473
Date:                Sun, 14 Mar 2021   Prob (F-statistic):              0.397
Time:                        06:03:16   Log-Likelihood:                 22.986
No. Observations:                  41   AIC:                            -39.97
Df Residuals:                      38   BIC:                            -34.83
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0059      0.048      0.124      0.902      -0.091       0.103
gprice         1.5665      1.139      1.375      0.177      -0.740       3.873
t           3.698e-05      0.002      0.019      0.985      -0.004       0.004
==============================================================================
Omnibus:                        1.185   Durbin-Watson:                   1.765
Prob(Omnibus):                  0.553   Jarque-Bera (JB):                0.463
Skew:                           0.197   Prob(JB):                        0.793
Kurtosis:                       3.340   Cond. No.                     1.27e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.27e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

C1.i The autocorrelation for $log(invpc)$ is 0.64, and after detrending is 0.49. The autocorrelation for $log(price)$ is 0.95 and 0.82 after detrending. By the more conservative test ($\hat{\rho}$ > 0.8), both versions of $log(price)$ suggest a unit root. By the $\hat{\rho} > 0.9$ test, only the series without detrending has the unit root (though we should still be cautious around $log(price)$.

C1.ii Estimates are reported above. With a log-log model the coefficient is a measure of price elasticity. Specifically, housing investment per capita increases by 3.88% for each 1% increase in growth of housing prices.

C1.iii $R^2$ falls from 0.510 to 0.303.

C1.iv All of the significance of the model is lost and $R^2$ now falls to 0.047. The estimate for $\Delta log(price)$ has fallen substantially. The time trend is no longer significant, though this should not be surprising since differencing should eliminate the time trend.

In [9]:
# Exercise 2
earns = pd.read_stata("stata/EARNS.DTA")
X = sm.add_constant(earns[["goutphr", "goutph_1"]])
sm.OLS(earns.ghrwage, X, missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                ghrwage   R-squared:                       0.493
Model:                            OLS   Adj. R-squared:                  0.465
Method:                 Least Squares   F-statistic:                     17.51
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           4.88e-06
Time:                        06:03:16   Log-Likelihood:                 108.63
No. Observations:                  39   AIC:                            -211.3
Df Residuals:                      36   BIC:                            -206.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0104      0.005     -2.294      0.028      -0.020      -0.001
goutphr        0.7284      0.167      4.356      0.000       0.389       1.068
goutph_1       0.4576      0.166      2.763      0.009       0.122       0.794
==============================================================================
Omnibus:                        1.236   Durbin-Watson:                   1.187
Prob(Omnibus):                  0.539   Jarque-Bera (JB):                0.639
Skew:                           0.303   Prob(JB):                        0.726
Kurtosis:                       3.162   Cond. No.                         75.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
earns["hint"] = earns.goutph_1 - earns.goutphr
X = sm.add_constant(earns[["goutphr", "hint"]])
model = sm.OLS(earns.ghrwage, X, missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                ghrwage   R-squared:                       0.493
Model:                            OLS   Adj. R-squared:                  0.465
Method:                 Least Squares   F-statistic:                     17.51
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           4.88e-06
Time:                        06:03:16   Log-Likelihood:                 108.63
No. Observations:                  39   AIC:                            -211.3
Df Residuals:                      36   BIC:                            -206.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0104      0.005     -2.294      0.028      -0.020      -0.001
goutphr        1.1860      0.203      5.838      0.000       0.774       1.598
hint           0.4576      0.166      2.763      0.009       0.122       0.794
==============================================================================
Omnibus:                        1.236   Durbin-Watson:                   1.187
Prob(Omnibus):                  0.539   Jarque-Bera (JB):                0.639
Skew:                           0.303   Prob(JB):                        0.726
Kurtosis:                       3.162   Cond. No.                         95.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
print("t statistic for for null of theta = 1:", (model.params[1] - 1) / model.bse[1])

t statistic for for null of theta = 1: 0.9156135280692587


In [12]:
X = sm.add_constant(earns[["goutphr", "goutph_1", "goutph_2"]])
sm.OLS(earns.ghrwage, X, missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                ghrwage   R-squared:                       0.515
Model:                            OLS   Adj. R-squared:                  0.472
Method:                 Least Squares   F-statistic:                     12.04
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.58e-05
Time:                        06:03:16   Log-Likelihood:                 107.97
No. Observations:                  38   AIC:                            -207.9
Df Residuals:                      34   BIC:                            -201.4
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0113      0.005     -2.331      0.026      -0.021      -0.001
goutphr        0.7464      0.162      4.622      0.000       0.418       1.075
goutph_1       0.3740      0.167      2.246      0.031       0.036       0.712
goutph_2       0.0653      0.160      0.409      0.685      -0.259       0.390
==============================================================================
Omnibus:                        1.512   Durbin-Watson:                   1.142
Prob(Omnibus):                  0.470   Jarque-Bera (JB):                0.768
Skew:                           0.324   Prob(JB):                        0.681
Kurtosis:                       3.256   Cond. No.                         77.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C2.i The lagged value is statistically significant at the 1% level

C2.ii Using the hint we estimate theta directly. The t statistic is 0.92 and so we have no evidence to reject the null hypothesis that $\beta_1 + \beta_2 = 1$

C2.iii Estimating the model with the second lag produces a small coefficient and p-value of 0.685. There does not appear to be any reason to include the additional lag.

In [13]:
# Exercise 3
nyse = pd.read_stata("stata/NYSE.DTA")
nyse["return_1_sq"] = nyse.return_1 ** 2

X = sm.add_constant(nyse[["return_1", "return_1_sq"]])
model = sm.OLS(nyse["return"], X, missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 return   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     2.160
Date:                Sun, 14 Mar 2021   Prob (F-statistic):              0.116
Time:                        06:03:17   Log-Likelihood:                -1490.3
No. Observations:                 689   AIC:                             2987.
Df Residuals:                     686   BIC:                             3000.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.2255      0.087      2.586      0.010       0.054       0.397
return_1        0.0486      0.039      1.254      0.210      -0.027       0.125
return_1_sq    -0.0097      0.007     -1.385      0.167      -0.024       0.004
==============================================================================
Omnibus:                      100.674   Durbin-Watson:                   1.988
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              577.296
Skew:                          -0.493   Prob(JB):                    4.38e-126
Kurtosis:                       7.375   Cond. No.                         13.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [14]:
model.f_test("return_1 = return_1_sq = 0") # This is also reported in the summary

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[2.16026656]]), p=0.11607808808035032, df_denom=686, df_num=2>

In [15]:
nyse["return_1Xreturn_2"] = nyse.return_1 * nyse.return_1.shift(1)

X = sm.add_constant(nyse[["return_1", "return_1Xreturn_2"]])
model = sm.OLS(nyse["return"], X, missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 return   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     1.802
Date:                Sun, 14 Mar 2021   Prob (F-statistic):              0.166
Time:                        06:03:17   Log-Likelihood:                -1488.9
No. Observations:                 688   AIC:                             2984.
Df Residuals:                     685   BIC:                             2997.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.1732      0.081      2.139      0.033       0.014       0.332
return_1              0.0687      0.039      1.751      0.080      -0.008       0.146
return_1Xreturn_2     0.0113      0.010      1.132      0.258      -0.008       0.031
==============================================================================
Omnibus:                      119.326   Durbin-Watson:                   1.996
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              672.910
Skew:                          -0.637   Prob(JB):                    7.58e-147
Kurtosis:                       7.675   Cond. No.                         8.36
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C3.i Results reported above

C3.ii The F-test is reported in the summary but directly carried out in the following cell. The F-test is not significant at the 10% level and so we have no evidence that the expected value of $return_t$ depends on $return_{t-1}$ (specifically, if both $\beta_1$ and $\beta_2$ are equal to zero, then the expectation does not depend on the the lagged term. One concern is that the null assumes the values are 0, and so we are assuming that $E(return_t | return_{t-1})$ does not depend on $return_{t-1}$).

C3.iii The test is similar as before. The individual parameters are not significant and the F-test is not significant, even at the 10% level. Again, we fail to reject the null that the coefficients are anything other than zero, and so past returns do not appear to affect present returns.

C3.iv From these models we find no evidence that weekly stock returns can be predicted based on past stock returns. The $R^2$ is small, and the F-tests did not provide us any reason to think previous returns had any effect on present returns.

In [16]:
# Exercise 4
phillips = pd.read_stata("stata/phillips.dta")
phillips = phillips[phillips.year < 1997]

sm.OLS(phillips.cinf, sm.add_constant(phillips.cunem), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cinf   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.116
Method:                 Least Squares   F-statistic:                     7.182
Date:                Sun, 14 Mar 2021   Prob (F-statistic):             0.0102
Time:                        06:03:17   Log-Likelihood:                -109.37
No. Observations:                  48   AIC:                             222.7
Df Residuals:                      46   BIC:                             226.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0782      0.348     -0.224      0.823      -0.780       0.623
cunem         -0.8422      0.314     -2.680      0.010      -1.475      -0.210
==============================================================================
Omnibus:                        8.000   Durbin-Watson:                   1.849
Prob(Omnibus):                  0.018   Jarque-Bera (JB):                8.518
Skew:                          -0.605   Prob(JB):                       0.0141
Kurtosis:                       4.671   Cond. No.                         1.11
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C4.i Results reported above. The effect for unemployment is large, negative, and statistically significant, more so than in the original model.

C4.ii From the $R^2$ (and adjusted $R^2$) the differenced model is a better fit.

In [17]:
# Exercise 5
fertil3 = pd.read_stata("stata/FERTIL3.DTA")
sm.OLS(fertil3.cgfr, sm.add_constant(fertil3[["cpe", "cpe_1", "cpe_2", "t"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cgfr   R-squared:                       0.234
Model:                            OLS   Adj. R-squared:                  0.186
Method:                 Least Squares   F-statistic:                     4.881
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00169
Time:                        06:03:17   Log-Likelihood:                -188.98
No. Observations:                  69   AIC:                             388.0
Df Residuals:                      64   BIC:                             399.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2674      1.046     -1.211      0.230      -3.358       0.823
cpe           -0.0348      0.027     -1.277      0.206      -0.089       0.020
cpe_1         -0.0131      0.028     -0.472      0.639      -0.069       0.043
cpe_2          0.1111      0.027      4.073      0.000       0.057       0.166
t              0.0079      0.024      0.325      0.746      -0.041       0.056
==============================================================================
Omnibus:                        4.498   Durbin-Watson:                   1.420
Prob(Omnibus):                  0.105   Jarque-Bera (JB):                4.016
Skew:                           0.354   Prob(JB):                        0.134
Kurtosis:                       3.946   Cond. No.                         96.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
model = sm.OLS(fertil3.cgfr, sm.add_constant(fertil3[["cpe", "cpe_1", "cpe_2", "ww2", "pill"]]), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cgfr   R-squared:                       0.296
Model:                            OLS   Adj. R-squared:                  0.240
Method:                 Least Squares   F-statistic:                     5.288
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           0.000406
Time:                        06:03:17   Log-Likelihood:                -186.07
No. Observations:                  69   AIC:                             384.1
Df Residuals:                      63   BIC:                             397.5
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.6503      0.582     -1.118      0.268      -1.813       0.512
cpe           -0.0752      0.032     -2.323      0.023      -0.140      -0.011
cpe_1         -0.0514      0.033     -1.550      0.126      -0.118       0.015
cpe_2          0.0883      0.028      3.155      0.002       0.032       0.144
ww2            4.8392      2.832      1.709      0.092      -0.820      10.498
pill          -1.6761      1.005     -1.668      0.100      -3.684       0.332
==============================================================================
Omnibus:                        0.463   Durbin-Watson:                   1.562
Prob(Omnibus):                  0.793   Jarque-Bera (JB):                0.112
Skew:                           0.065   Prob(JB):                        0.945
Kurtosis:                       3.149   Cond. No.                         129.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [19]:
model.f_test("ww2 = pill = 0")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[2.8230719]]), p=0.06696050459945173, df_denom=63, df_num=2>

In [20]:
sm.OLS(fertil3.cgfr, sm.add_constant(fertil3[["cpe", "cpe_1", "cpe_2", "ww2", "pill", "t"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cgfr   R-squared:                       0.359
Model:                            OLS   Adj. R-squared:                  0.297
Method:                 Least Squares   F-statistic:                     5.792
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           7.77e-05
Time:                        06:03:17   Log-Likelihood:                -182.81
No. Observations:                  69   AIC:                             379.6
Df Residuals:                      62   BIC:                             395.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.1412      1.150     -2.732      0.008      -5.439      -0.843
cpe           -0.0618      0.032     -1.958      0.055      -0.125       0.001
cpe_1         -0.0391      0.032     -1.213      0.230      -0.104       0.025
cpe_2          0.0952      0.027      3.519      0.001       0.041       0.149
ww2            3.2508      2.797      1.162      0.250      -2.341       8.842
pill          -4.8881      1.616     -3.025      0.004      -8.118      -1.658
t              0.0944      0.038      2.480      0.016       0.018       0.170
==============================================================================
Omnibus:                        1.213   Durbin-Watson:                   1.729
Prob(Omnibus):                  0.545   Jarque-Bera (JB):                0.800
Skew:                           0.258   Prob(JB):                        0.670
Kurtosis:                       3.112   Cond. No.                         279.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [21]:
fertil3["pe_1-pe"] = fertil3.cpe_1 - fertil3.cpe
fertil3["pe_2-pe"] = fertil3.cpe_2 - fertil3.cpe
sm.OLS(fertil3.cgfr, sm.add_constant(fertil3[["cpe", "pe_1-pe", "pe_2-pe", "ww2", "pill", "t"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cgfr   R-squared:                       0.359
Model:                            OLS   Adj. R-squared:                  0.297
Method:                 Least Squares   F-statistic:                     5.792
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           7.77e-05
Time:                        06:03:17   Log-Likelihood:                -182.81
No. Observations:                  69   AIC:                             379.6
Df Residuals:                      62   BIC:                             395.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -3.1412      1.150     -2.732      0.008      -5.439      -0.843
cpe           -0.0058      0.061     -0.094      0.925      -0.128       0.117
pe_1-pe       -0.0391      0.032     -1.213      0.230      -0.104       0.025
pe_2-pe        0.0952      0.027      3.519      0.001       0.041       0.149
ww2            3.2508      2.797      1.162      0.250      -2.341       8.842
pill          -4.8881      1.616     -3.025      0.004      -8.118      -1.658
t              0.0944      0.038      2.480      0.016       0.018       0.170
==============================================================================
Omnibus:                        1.213   Durbin-Watson:                   1.729
Prob(Omnibus):                  0.545   Jarque-Bera (JB):                0.800
Skew:                           0.258   Prob(JB):                        0.670
Kurtosis:                       3.112   Cond. No.                         279.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C5.i The time trend is not statistically significant. The difference should take care of the time effects so this should not be particularly surprising.

C5.ii The ww2 and pill variables are not jointly significant at the 5% level, though they are at the 10%.

C5.iii The time trend become larger and is now statistically significant (this was surprising to me since differencing should account for time trends). The coefficient for pill becomes larger and statistically significant.

C5.iv Using what we learned from chapter 10, we regress with subtracting $cpe$ from the lags. The coefficient for cpe reports the LRP and reports a standard error. Compared to 10.19, the LRP is now small, negative, and not statistically significant and so the relationship between fertility and the personal exemption does not seem to be a particularly robust finding.

In [22]:
# Exercise 6
inven = pd.read_stata("stata/INVEN.DTA")
sm.OLS(inven.cinven, sm.add_constant(inven.cgdp), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 cinven   R-squared:                       0.554
Model:                            OLS   Adj. R-squared:                  0.540
Method:                 Least Squares   F-statistic:                     42.16
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.99e-07
Time:                        06:03:17   Log-Likelihood:                -141.72
No. Observations:                  36   AIC:                             287.4
Df Residuals:                      34   BIC:                             290.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.5888      3.641      0.711      0.482      -4.810       9.988
cgdp           0.1525      0.023      6.493      0.000       0.105       0.200
==============================================================================
Omnibus:                        4.258   Durbin-Watson:                   2.197
Prob(Omnibus):                  0.119   Jarque-Bera (JB):                2.922
Skew:                          -0.489   Prob(JB):                        0.232
Kurtosis:                       3.996   Cond. No.                         265.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [23]:
sm.OLS(inven.cinven, sm.add_constant(inven[["cgdp", "r3"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 cinven   R-squared:                       0.562
Model:                            OLS   Adj. R-squared:                  0.536
Method:                 Least Squares   F-statistic:                     21.20
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.20e-06
Time:                        06:03:17   Log-Likelihood:                -141.37
No. Observations:                  36   AIC:                             288.7
Df Residuals:                      33   BIC:                             293.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.0042      3.694      0.813      0.422      -4.512      10.521
cgdp           0.1590      0.025      6.376      0.000       0.108       0.210
r3            -0.8953      1.101     -0.813      0.422      -3.135       1.344
==============================================================================
Omnibus:                        2.128   Durbin-Watson:                   2.272
Prob(Omnibus):                  0.345   Jarque-Bera (JB):                1.093
Skew:                          -0.350   Prob(JB):                        0.579
Kurtosis:                       3.489   Cond. No.                         268.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [24]:
sm.OLS(inven.cinven, sm.add_constant(inven[["cgdp", "cr3"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 cinven   R-squared:                       0.555
Model:                            OLS   Adj. R-squared:                  0.528
Method:                 Least Squares   F-statistic:                     20.56
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.59e-06
Time:                        06:03:17   Log-Likelihood:                -141.67
No. Observations:                  36   AIC:                             289.3
Df Residuals:                      33   BIC:                             294.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.3362      3.782      0.618      0.541      -5.358      10.031
cgdp           0.1545      0.025      6.255      0.000       0.104       0.205
cr3           -0.4697      1.540     -0.305      0.762      -3.602       2.663
==============================================================================
Omnibus:                        4.034   Durbin-Watson:                   2.192
Prob(Omnibus):                  0.133   Jarque-Bera (JB):                2.683
Skew:                          -0.487   Prob(JB):                        0.261
Kurtosis:                       3.917   Cond. No.                         273.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C6.i Results reported above. $\hat{\beta_1}$ is statistically significant at the 1% level even for a two sided test (and the coefficient is positive).

C6.ii The level for the interest rate is negative as we might expect (as the text describes, an increase in interest rates increases the opportunity cost for holding inventories and so should decrease), but the t statistic is small and so we cannot conclude it actually affects inventories.

C6.iii When using the first difference the magnitude decreaess but remains negative. The effect is still large and backed by economic theory, but the confidence interval is enormous and includes 0 and so we still cannot conclude that interest rates affect inventories.

In [25]:
# Exercise 7
consump = pd.read_stata("stata/consump.dta")
sm.OLS(consump.gc, sm.add_constant(consump.gc_1), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     gc   R-squared:                       0.199
Model:                            OLS   Adj. R-squared:                  0.174
Method:                 Least Squares   F-statistic:                     8.174
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00731
Time:                        06:03:17   Log-Likelihood:                 107.34
No. Observations:                  35   AIC:                            -210.7
Df Residuals:                      33   BIC:                            -207.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0114      0.004      3.026      0.005       0.004       0.019
gc_1           0.4461      0.156      2.859      0.007       0.129       0.764
==============================================================================
Omnibus:                        1.381   Durbin-Watson:                   1.927
Prob(Omnibus):                  0.501   Jarque-Bera (JB):                0.754
Skew:                          -0.353   Prob(JB):                        0.686
Kurtosis:                       3.134   Cond. No.                         79.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [26]:
consump["i3_1"] = consump.i3.shift(1)
consump["inf_1"] = consump.inf.shift(1)
model = sm.OLS(consump.gc, sm.add_constant(consump[["gc_1", "gy_1", "i3_1", "inf_1"]]), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     gc   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.211
Method:                 Least Squares   F-statistic:                     3.273
Date:                Sun, 14 Mar 2021   Prob (F-statistic):             0.0243
Time:                        06:03:17   Log-Likelihood:                 109.80
No. Observations:                  35   AIC:                            -209.6
Df Residuals:                      30   BIC:                            -201.8
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0226      0.007      3.187      0.003       0.008       0.037
gc_1           0.4336      0.290      1.497      0.145      -0.158       1.025
gy_1          -0.1079      0.195     -0.554      0.583      -0.505       0.290
i3_1          -0.0007      0.001     -0.672      0.507      -0.003       0.002
inf_1         -0.0008      0.001     -0.825      0.416      -0.003       0.001
==============================================================================
Omnibus:                        0.118   Durbin-Watson:                   1.783
Prob(Omnibus):                  0.943   Jarque-Bera (JB):                0.163
Skew:                          -0.117   Prob(JB):                        0.922
Kurtosis:                       2.762   Cond. No.                     1.52e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.52e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [27]:
model.f_test("gy_1 = i3_1 = inf_1 = 0")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[1.51224519]]), p=0.23147449530935482, df_denom=30, df_num=3>

C7.i Results reported above. The null hypothesis is that current information, in this case the present value for gc (growth in real per capita consumption), does not depend the past value for gc, or in the terms of our model that $\beta_1 = 0$. The alternative hypothesis is that it does, (specifically $E(gc_t | I_{t-1} \neq E(gc_t)$) meaning that $\beta_1 \neq 0$.

The $\hat{\beta_1}$ is positive and significant, rejecting the null hypothesis and thus the permanent income hypothesis (PIH).

C7.ii Results reported above. The added values are not individually significant or jointly significant.

C7.iii $\hat{\beta_1}$ is now no longer individually significant and so we might be tempted to accept the PIH, it is important to rember that the PIH relies on $I_{t-1}$ which would imply a test of the joint significance for all values composing $I$. This is dealt with in the next question. We are no longer testing the same hypothesis as in (i) and so it is not sufficient to look at the individual significance of $gc_1$

C7.iv The F statistic for the joint significance of the model is 3.273 (p-value 0.024). We still reject the PIH given that $gc$ does depend on $I_{t-1}$ even if we cannot identify an individual component. The joint significance is the appropriate test and rejects the null, albeit now at the 5% level.

In [28]:
# Exercise 8
phillips = pd.read_stata("stata/phillips.dta")
model = sm.OLS(phillips.unem, sm.add_constant(phillips.unem_1), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   unem   R-squared:                       0.566
Model:                            OLS   Adj. R-squared:                  0.558
Method:                 Least Squares   F-statistic:                     69.12
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           3.54e-11
Time:                        06:03:17   Log-Likelihood:                -76.946
No. Observations:                  55   AIC:                             157.9
Df Residuals:                      53   BIC:                             161.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.4897      0.520      2.864      0.006       0.446       2.533
unem_1         0.7424      0.089      8.314      0.000       0.563       0.921
==============================================================================
Omnibus:                        9.490   Durbin-Watson:                   1.666
Prob(Omnibus):                  0.009   Jarque-Bera (JB):                8.954
Skew:                           0.922   Prob(JB):                       0.0114
Kurtosis:                       3.714   Cond. No.                         23.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [29]:
model.params.dot(np.array([1, 6])) # (6 is unemployment rate from 2003)

5.943978889415067

In [30]:
model = sm.OLS(phillips.unem, sm.add_constant(phillips[["unem_1", "inf_1"]]), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   unem   R-squared:                       0.696
Model:                            OLS   Adj. R-squared:                  0.685
Method:                 Least Squares   F-statistic:                     59.60
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           3.51e-14
Time:                        06:03:17   Log-Likelihood:                -67.132
No. Observations:                  55   AIC:                             140.3
Df Residuals:                      52   BIC:                             146.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2964      0.441      2.938      0.005       0.411       2.182
unem_1         0.6495      0.078      8.333      0.000       0.493       0.806
inf_1          0.1830      0.039      4.722      0.000       0.105       0.261
==============================================================================
Omnibus:                        4.791   Durbin-Watson:                   1.806
Prob(Omnibus):                  0.091   Jarque-Bera (JB):                3.743
Skew:                           0.578   Prob(JB):                        0.154
Kurtosis:                       3.545   Cond. No.                         28.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [31]:
model.params.dot(np.array([1, 6, 2.3])) # Inflation was 2.3

5.6142082338351535

In [32]:
model = sm.OLS(phillips.unem, sm.add_constant(pd.DataFrame([phillips.unem_1 - 6, phillips.inf_1 - 2.3]).T), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   unem   R-squared:                       0.696
Model:                            OLS   Adj. R-squared:                  0.685
Method:                 Least Squares   F-statistic:                     59.60
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           3.51e-14
Time:                        06:03:17   Log-Likelihood:                -67.132
No. Observations:                  55   AIC:                             140.3
Df Residuals:                      52   BIC:                             146.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.6142      0.136     41.166      0.000       5.341       5.888
unem_1         0.6495      0.078      8.333      0.000       0.493       0.806
inf_1          0.1830      0.039      4.722      0.000       0.105       0.261
==============================================================================
Omnibus:                        4.791   Durbin-Watson:                   1.806
Prob(Omnibus):                  0.091   Jarque-Bera (JB):                3.743
Skew:                           0.578   Prob(JB):                        0.154
Kurtosis:                       3.545   Cond. No.                         4.31
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [33]:
se_y = np.sqrt((model.bse[0] ** 2) + model.mse_resid)

print("Confidence Interval:",model.params[0] - (1.96 * se_y), ",", model.params[0] + (1.96 * se_y))

Confidence Interval: 3.939674782507841 , 7.288741675700037


C8.i Results above. The estimated value is 5.94 which is higher than the true value of 5.5.

C8.ii The lagged inflation is statistically significant.

C8.iii The predicted inflation is 5.61, which is closer to the true value (5.5) than in C8.i

C8.iv Using the method from 6.4 the confidence interval is 3.94 to 7.29 which includes the true value.

In [34]:
# Exercise 9
traffic2 = pd.read_stata("stata/TRAFFIC2.DTA")
traffic2.prcfat.corr(traffic2.prcfat_1)

0.7086404674144829

In [35]:
traffic2.unem.corr(traffic2.unem.shift(1))

0.9497837189727667

In [36]:
traffic2["cprcfat"] = traffic2.prcfat - traffic2.prcfat_1
traffic2["cunem"] = traffic2.unem - traffic2.unem.shift(1)

sm.OLS(traffic2.cprcfat, sm.add_constant(traffic2[["cunem", "spdlaw", "beltlaw", "wkends", "t", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                cprcfat   R-squared:                       0.351
Model:                            OLS   Adj. R-squared:                  0.224
Method:                 Least Squares   F-statistic:                     2.773
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00136
Time:                        06:03:18   Log-Likelihood:                 138.47
No. Observations:                  99   AIC:                            -242.9
Df Residuals:                      82   BIC:                            -198.8
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.1148      0.110     -1.043      0.300      -0.334       0.104
cunem          0.0128      0.016      0.785      0.435      -0.020       0.045
spdlaw        -0.0076      0.023     -0.322      0.748      -0.054       0.039
beltlaw        0.0013      0.026      0.048      0.961      -0.051       0.054
wkends         0.0060      0.008      0.785      0.435      -0.009       0.021
t              0.0001      0.000      0.260      0.796      -0.001       0.001
feb            0.0334      0.037      0.908      0.367      -0.040       0.107
mar            0.0418      0.039      1.085      0.281      -0.035       0.119
apr            0.0983      0.038      2.590      0.011       0.023       0.174
may            0.0562      0.038      1.474      0.144      -0.020       0.132
jun            0.0681      0.035      1.934      0.057      -0.002       0.138
jul            0.0669      0.033      2.011      0.048       0.001       0.133
aug            0.0658      0.039      1.672      0.098      -0.012       0.144
sep            0.0086      0.039      0.218      0.828      -0.070       0.087
oct           -0.0306      0.035     -0.872      0.386      -0.100       0.039
nov           -0.0658      0.036     -1.841      0.069      -0.137       0.005
dec            0.0271      0.037      0.734      0.465      -0.046       0.100
==============================================================================
Omnibus:                        1.082   Durbin-Watson:                   2.536
Prob(Omnibus):                  0.582   Jarque-Bera (JB):                1.045
Skew:                           0.096   Prob(JB):                        0.593
Kurtosis:                       2.535   Cond. No.                     1.14e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.14e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

C9.i There does not appear to be autocorrelation for prcfat but the correlation between the current and lagged unem is almost 0.95, suggesting autocorrelation.

C9.ii Results reported above. The F-test for the overall model shows significance, but there are no noteworthy individual effects (only the April dummy is significant).

C9.iii The results clearly are not similar to to the results using levels. By first differencing we have lost any potentially interesting effects. Of course, in the presence of autocorrelation, the interesting effects we find through the levels may simply be incorrect.

In [37]:
# Exercise 10
model = sm.OLS(phillips.cinf, sm.add_constant(phillips.unem), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cinf   R-squared:                       0.104
Model:                            OLS   Adj. R-squared:                  0.087
Method:                 Least Squares   F-statistic:                     6.132
Date:                Sun, 14 Mar 2021   Prob (F-statistic):             0.0165
Time:                        06:03:18   Log-Likelihood:                -123.00
No. Observations:                  55   AIC:                             250.0
Df Residuals:                      53   BIC:                             254.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.8282      1.225      2.309      0.025       0.371       5.285
unem          -0.5176      0.209     -2.476      0.017      -0.937      -0.098
==============================================================================
Omnibus:                       25.286   Durbin-Watson:                   1.771
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               71.207
Skew:                          -1.193   Prob(JB):                     3.45e-16
Kurtosis:                       8.038   Cond. No.                         23.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [38]:
model.params[0] / (-model.params[1])

5.463554363192442

In [39]:
phillips.unem.corr(phillips.unem_1)

0.7523375261763339

In [40]:
sm.OLS(phillips.cinf, sm.add_constant(phillips.cunem), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   cinf   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.118
Method:                 Least Squares   F-statistic:                     8.256
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00583
Time:                        06:03:18   Log-Likelihood:                -122.03
No. Observations:                  55   AIC:                             248.1
Df Residuals:                      53   BIC:                             252.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0721      0.306     -0.236      0.814      -0.686       0.541
cunem         -0.8328      0.290     -2.873      0.006      -1.414      -0.251
==============================================================================
Omnibus:                       10.746   Durbin-Watson:                   1.841
Prob(Omnibus):                  0.005   Jarque-Bera (JB):               15.025
Skew:                          -0.644   Prob(JB):                     0.000546
Kurtosis:                       5.213   Cond. No.                         1.07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C10.i Results reported above. The intercept and coefficient do not seem to change substantially.

C10.ii The estimate for the natural rate of unemplayment is slightly lower at 5.46 (the example was 5.58).

C10.iii The first order autocorrelation is 0.75. This is below even the more conservative test  from the chapter and so we would not think there a unit root (by these standards).

C10.iv The differenced model gives the higher $R^2$ and cunem has a smaller p-value.

In [41]:
# Exercise 11
okun = pd.read_stata("stata/okun.dta")

model = sm.OLS(okun.pcrgdp, sm.add_constant(okun.cunem), missing = "drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 pcrgdp   R-squared:                       0.710
Model:                            OLS   Adj. R-squared:                  0.704
Method:                 Least Squares   F-statistic:                     107.9
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           2.04e-13
Time:                        06:03:18   Log-Likelihood:                -68.769
No. Observations:                  46   AIC:                             141.5
Df Residuals:                      44   BIC:                             145.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.3444      0.163     20.559      0.000       3.017       3.672
cunem         -1.8909      0.182    -10.388      0.000      -2.258      -1.524
==============================================================================
Omnibus:                        0.827   Durbin-Watson:                   1.857
Prob(Omnibus):                  0.661   Jarque-Bera (JB):                0.791
Skew:                           0.033   Prob(JB):                        0.674
Kurtosis:                       2.361   Cond. No.                         1.12
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [42]:
(model.params[1] + 2) / model.bse[1]

0.5992879191676945

In [43]:
(model.params[0] - 3) / model.bse[0]

2.1172785700408667

In [44]:
model.f_test("const = 3, cunem = -2")

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=array([[2.40888936]]), p=0.10168164461818073, df_denom=44, df_num=2>

C11.i The point estimates are not exactly the anticipated values, but it would be highly unusual if they were. The 95% confidence interval for the constant is (3.017, 3.672) and so falls above the anticipated value of 3. The coefficient for the change in unemployment is  (-2.258, -1.524) which includes the anticipated value of -2. The biggest rejection for the model at this point would be on the basis of the constant, which is still close.

C11.ii The t statistic for testing if $\hat{\beta_1} = -2$ is about 0.6. We do not reject the null against any reasonable significance level. This is not particularly surprising seeing as -2 feel quite comfortably in our 95% confidence interval.

C11.iii The t statistic for the test that $\hat{\beta_0} = 3$ is 2.117. This is sufficient to reject at the 5% level (again, unsurprising based on our confidence interval). We should decide a standard and stick to it and so rejection is a rejection, it certainly is on the border and we could not reject at the 1% level.

C11.iv The F test gives a value of 2.41 with a p value of about 0.101. Given that the null hypothesis is Okun's law we fail to reject it though if we are focusing at the 10% level it is right at the threshold. The data tends to support Okun's law (or at least fails to provide sufficient evidence that it is different).

In [45]:
# Exercise 12
minwage = pd.read_stata("stata/minwage.dta")
minwage["gwage232_1"] = minwage.gwage232.shift(1)
minwage["gemp232_1"] = minwage.gemp232.shift(1)
minwage.gwage232.corr(minwage.gwage232_1)

-0.03494393406515538

In [46]:
sm.OLS(minwage.gwage232, sm.add_constant(minwage[["gwage232_1", "gmwage", "gcpi"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               gwage232   R-squared:                       0.299
Model:                            OLS   Adj. R-squared:                  0.295
Method:                 Least Squares   F-statistic:                     85.99
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           2.31e-46
Time:                        06:03:18   Log-Likelihood:                 2090.3
No. Observations:                 610   AIC:                            -4173.
Df Residuals:                     606   BIC:                            -4155.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0024      0.000      5.572      0.000       0.002       0.003
gwage232_1    -0.0779      0.034     -2.272      0.023      -0.145      -0.011
gmwage         0.1518      0.010     15.738      0.000       0.133       0.171
gcpi           0.2631      0.082      3.191      0.001       0.101       0.425
==============================================================================
Omnibus:                       75.970   Durbin-Watson:                   2.027
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              322.973
Skew:                           0.476   Prob(JB):                     7.37e-71
Kurtosis:                       6.435   Cond. No.                         258.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [47]:
sm.OLS(minwage.gwage232, sm.add_constant(minwage[["gwage232_1", "gmwage", "gcpi", "gemp232_1"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               gwage232   R-squared:                       0.316
Model:                            OLS   Adj. R-squared:                  0.311
Method:                 Least Squares   F-statistic:                     69.80
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.36e-48
Time:                        06:03:18   Log-Likelihood:                 2097.8
No. Observations:                 610   AIC:                            -4186.
Df Residuals:                     605   BIC:                            -4164.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0025      0.000      5.753      0.000       0.002       0.003
gwage232_1    -0.0745      0.034     -2.199      0.028      -0.141      -0.008
gmwage         0.1527      0.010     16.007      0.000       0.134       0.171
gcpi           0.2523      0.082      3.094      0.002       0.092       0.412
gemp232_1      0.0661      0.017      3.899      0.000       0.033       0.099
==============================================================================
Omnibus:                       76.851   Durbin-Watson:                   1.975
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              325.844
Skew:                           0.485   Prob(JB):                     1.75e-71
Kurtosis:                       6.447   Cond. No.                         259.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [48]:
sm.OLS(minwage.gwage232, sm.add_constant(minwage[["gmwage", "gcpi"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               gwage232   R-squared:                       0.293
Model:                            OLS   Adj. R-squared:                  0.290
Method:                 Least Squares   F-statistic:                     125.8
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.91e-46
Time:                        06:03:18   Log-Likelihood:                 2091.6
No. Observations:                 611   AIC:                            -4177.
Df Residuals:                     608   BIC:                            -4164.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0022      0.000      5.185      0.000       0.001       0.003
gmwage         0.1506      0.010     15.591      0.000       0.132       0.170
gcpi           0.2435      0.082      2.965      0.003       0.082       0.405
==============================================================================
Omnibus:                       73.109   Durbin-Watson:                   2.194
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              306.495
Skew:                           0.455   Prob(JB):                     2.79e-67
Kurtosis:                       6.348   Cond. No.                         257.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [49]:
sm.OLS(minwage.gwage232, sm.add_constant(minwage[["gwage232_1", "gemp232_1"]]), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               gwage232   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                  0.013
Method:                 Least Squares   F-statistic:                     4.979
Date:                Sun, 14 Mar 2021   Prob (F-statistic):            0.00717
Time:                        06:03:18   Log-Likelihood:                 1987.1
No. Observations:                 610   AIC:                            -3968.
Df Residuals:                     607   BIC:                            -3955.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0038      0.000      9.298      0.000       0.003       0.005
gwage232_1    -0.0321      0.040     -0.797      0.426      -0.111       0.047
gemp232_1      0.0616      0.020      3.034      0.003       0.022       0.101
==============================================================================
Omnibus:                      344.245   Durbin-Watson:                   1.953
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3840.128
Skew:                           2.283   Prob(JB):                         0.00
Kurtosis:                      14.412   Cond. No.                         107.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C12.i The correlation is -0.035. We should be reasonably confident that the series is weakly dependent.

C12.ii From the model we see that a 1% increase in the federal minimum wage is expected to increase the wages in the sector by about 0.152 percent (with previous period wage and cpi growth held constant). The result is statistically significant.

C12.iii When adding lagged employment growth to the model it is statistically significant.

C12.iv After estimating the model without the lags and comparing the two, the result remains fairly consistent at 0.15 (and significant).

C12.v The $R^2$ of the regression of $gmwage_t$ on $gmwage232_{t-1}$ and $gemp232_{t-1}$ is 0.016. The lagged variables do very little to explain the variation in contemporary growth of wages and so this explains why so little changes over the different models.

In [50]:
# Exercise 13
beveridge = pd.read_stata("stata/beveridge.dta")
beveridge.urate.corr(beveridge.urate_1)

0.9957539715993459

In [51]:
beveridge.vrate.corr(beveridge.vrate_1)

0.9445816187674836

In [52]:
sm.OLS(beveridge.urate, sm.add_constant(beveridge.vrate)).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  urate   R-squared:                       0.711
Model:                            OLS   Adj. R-squared:                  0.709
Method:                 Least Squares   F-statistic:                     327.3
Date:                Sun, 14 Mar 2021   Prob (F-statistic):           1.14e-37
Time:                        06:03:18   Log-Likelihood:                -215.35
No. Observations:                 135   AIC:                             434.7
Df Residuals:                     133   BIC:                             440.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.1194      0.592     28.918      0.000      15.948      18.290
vrate         -3.7414      0.207    -18.091      0.000      -4.151      -3.332
==============================================================================
Omnibus:                        8.603   Durbin-Watson:                   0.293
Prob(Omnibus):                  0.014   Jarque-Bera (JB):                8.583
Skew:                           0.572   Prob(JB):                       0.0137
Kurtosis:                       2.536   Cond. No.                         18.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [53]:
sm.OLS(beveridge.curate, sm.add_constant(beveridge.cvrate), missing = "drop").fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 curate   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.007
Method:                 Least Squares   F-statistic:                   0.06614
Date:                Sun, 14 Mar 2021   Prob (F-statistic):              0.797
Time:                        06:03:18   Log-Likelihood:                 22.704
No. Observations:                 134   AIC:                            -41.41
Df Residuals:                     132   BIC:                            -35.61
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0370      0.018      2.080      0.039       0.002       0.072
cvrate        -0.0276      0.107     -0.257      0.797      -0.240       0.185
==============================================================================
Omnibus:                       11.363   Durbin-Watson:                   1.303
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               12.935
Skew:                           0.564   Prob(JB):                      0.00155
Kurtosis:                       4.022   Cond. No.                         6.04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

C13.i $urate$ and its lag are highly correlated (1 when rounding, 0.995). This is very likely a unit root process.

C13.ii $vrate$ and its lag are also likely a unit root process given the correlation of about 0.945. This is also likely a unit root process. It is quite unlikely we are dealing with weakly dependent processes.

C13.iii The results are reported above. $\hat{\beta_1} < 0$ as expected from theory. The results are statistically significant, but we should not ignore the findings of part (i) and (ii).

C13.iv Our first assumption is that we are dealing with a stochastic process that is linear, stationary, and weakly dependent. The presence of unit roots violates this assumption. If the model satisfies the assumptions in the previous chapter (chapter 10), we don't need to worry, but this is a stronger set of assumptions and is not likely to be the case.

C13.v When taking the difference the slope falls from -3 to almost 0. While the result still is negative, it is not statistically significant (with a t statistic of -0.257 and so we will not reject the null at any credible significance level). As before, the relationship may exist, but we cannot say this from the regression with the differences.